File: seed_progress.js
const { createClient } = require('@supabase/supabase-js');
const fs = require('fs');

// Read env variables (mocking process.env for simplicity, pulling directly from .env)
const env = fs.readFileSync('.env', 'utf-8').split('\n');
let SUPABASE_URL = '';
let SUPABASE_ANON_KEY = '';

env.forEach(line => {
    if (line.startsWith('SUPABASE_URL=')) SUPABASE_URL = line.split('=')[1].trim();
    if (line.startsWith('SUPABASE_ANON_KEY=')) SUPABASE_ANON_KEY = line.split('=')[1].trim();
});

const supabase = createClient(SUPABASE_URL, SUPABASE_ANON_KEY);

async function seedProgress() {
    console.log("Fetching courses...");
    const { data: courses, error: errC } = await supabase.from('courses').select('id');
    if (errC || !courses.length) return console.error("No courses found:", errC);

    console.log("Fetching students...");
    const { data: students, error: errS } = await supabase.from('users').select('id').eq('role', 'student');
    if (errS || !students.length) return console.error("No students found:", errS);

    console.log(`Seeding progress for ${students.length} students...`);
    
    let progressRecords = [];
    
    for (const student of students) {
        // Enrolled courses (3 to 5 courses)
        const shuffled = courses.sort(() => 0.5 - Math.random());
        const enrolled = shuffled.slice(0, Math.floor(Math.random() * 3) + 3);
        
        for (const course of enrolled) {
            const r = Math.random();
            let hours, modIdx, lessIdx, compLessons;

            if (r < 0.2) {
                // Completed
                hours = parseFloat((Math.random() * 10 + 10).toFixed(2));
                modIdx = 3; lessIdx = 0;
                compLessons = [
                    "00000000-0000-0000-0000-000000000001",
                    "00000000-0000-0000-0000-000000000002"
                ];
            } else if (r < 0.7) {
                // Active
                hours = parseFloat((Math.random() * 8 + 1).toFixed(2));
                modIdx = 1; lessIdx = 1;
                compLessons = ["00000000-0000-0000-0000-000000000001"];
            } else {
                // New
                hours = 0.0;
                modIdx = 0; lessIdx = 0;
                compLessons = [];
            }
            
            progressRecords.push({
                user_id: student.id,
                course_id: course.id,
                completed_lessons: compLessons,
                hours_spent: hours,
                current_module_index: modIdx,
                current_lesson_index: lessIdx
            });
        }
    }
    
    console.log(`Inserting ${progressRecords.length} records...`);
    // Insert all
    const { error } = await supabase.from('progress').upsert(progressRecords);
    if (error) {
        console.error("Insert failed:", error);
    } else {
        console.log("Progress seeded successfully!");
    }
}

seedProgress();

File: backend/scripts/seed_progress.py
import os
import sys
import json
import random
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from app.core.config import settings
from app.database.supabase_manager import get_supabase_client

def seed_ecosystem():
    supabase = get_supabase_client()
    
    print("Fetching existing courses...")
    courses = supabase.table('courses').select('id').execute().data
    if not courses:
        print("No courses found. Cannot seed enrollments/progress.")
        return
        
    print("Fetching existing students...")
    students = supabase.table('users').select('id').eq('role', 'student').execute().data
    
    if not students:
        print("No students found. Run basic Python user seed first if needed.")
        return
    
    print(f"Seeding progress for {len(students)} students across {len(courses)} courses...")
    
    progress_records = []
    
    for student in students:
        student_id = student['id']
        # Each student takes 3-5 random courses
        enrolled_courses = random.sample(courses, min(len(courses), random.randint(3, 5)))
        
        for course in enrolled_courses:
            course_id = course['id']
            r = random.random()
            
            if r < 0.2:
                # 20% Completed
                hours = round(random.uniform(10.0, 20.0), 2)
                mod_idx = 3
                less_idx = 0
                comp_lessons = [str(random.randint(1000, 9999)) for _ in range(5)]
            elif r < 0.7:
                # 50% Active
                hours = round(random.uniform(1.0, 9.0), 2)
                mod_idx = 1
                less_idx = 1
                comp_lessons = [str(random.randint(1000, 9999)) for _ in range(2)]
            else:
                # 30% New
                hours = 0.0
                mod_idx = 0
                less_idx = 0
                comp_lessons = []
                
            progress_records.append({
                'user_id': student_id,
                'course_id': course_id,
                'completed_lessons': comp_lessons,
                'hours_spent': hours,
                'current_module_index': mod_idx,
                'current_lesson_index': less_idx
            })
            
    # Batch insert to avoid huge payload limits
    batch_size = 50
    for i in range(0, len(progress_records), batch_size):
        batch = progress_records[i:i+batch_size]
        try:
           supabase.table('progress').upsert(batch).execute()
        except Exception as e:
           print(f"Error on batch {i}: {e}")
           
    print("Finished seeding progress via Supabase REST client.")

if __name__ == "__main__":
    seed_ecosystem()
